File: backend/finance/src/main/java/com/ngo/finance/donor/mapper/GrantMapper.java
package com.ngo.finance.donor.mapper;

import com.ngo.finance.donor.dto.request.CreateGrantRequest;
import com.ngo.finance.donor.dto.response.GrantDetailsResponse;
import com.ngo.finance.donor.dto.response.GrantListResponse;
import com.ngo.finance.donor.entity.GrantAgreement;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

/**
 * MapStruct mapper for Grant Agreement entity to/from DTOs
 */
@Mapper(componentModel = "spring")
public interface GrantMapper {

    @Mapping(target = "id", ignore = true)
    @Mapping(target = "donor", ignore = true)
    @Mapping(target = "fundProfile", ignore = true)
    @Mapping(target = "createdAt", ignore = true)
    @Mapping(target = "updatedAt", ignore = true)
    @Mapping(target = "createdBy", ignore = true)
    @Mapping(target = "updatedBy", ignore = true)
    @Mapping(target = "isActive", ignore = true)
    // Derived from the fund profile's tranche plan, never from the request.
    @Mapping(target = "totalGrantAmount", ignore = true)
    // Status and the approval block are applied by the service (they need
    // enum/date coercion and must not be silently defaulted by the mapper).
    @Mapping(target = "grantStatus", ignore = true)
    @Mapping(target = "isApproved", ignore = true)
    @Mapping(target = "approvedBy", ignore = true)
    @Mapping(target = "approvalDate", ignore = true)
    @Mapping(target = "approvalRemarks", ignore = true)
    GrantAgreement toEntity(CreateGrantRequest request);

    @Mapping(source = "grantStatus", target = "status")
    @Mapping(target = "approvedByName", ignore = true)
    @Mapping(source = "donor.id", target = "donorId")
    @Mapping(source = "donor.donorName", target = "donorName")
    @Mapping(source = "fundProfile.id", target = "fundProfileId")
    @Mapping(source = "fundProfile.fundClass", target = "fundClassCode")
    GrantDetailsResponse toDetailsResponse(GrantAgreement entity);

    @Mapping(source = "grantStatus", target = "status")
    @Mapping(source = "donor.donorName", target = "donorName")
    @Mapping(source = "fundProfile.id", target = "fundProfileId")
    @Mapping(source = "fundProfile.fundClass", target = "fundClassCode")
    GrantListResponse toListResponse(GrantAgreement entity);
}

File: frontend/src/features/donor-management/mappers/grantMapper.js
/**
 * GrantMapper — GrantListResponse / GrantDetailsResponse ↔ view models,
 * form values → CreateGrantRequest. Field names mirror the backend DTOs.
 *
 * A grant's class is the profile's restriction class (A/B/C, `fundClassCode`),
 * not the donor typology enum.
 */

/** Agreement status (section 1 of the form) → display label. */
export const GRANT_STATUS_LABEL = {
  ACTIVE: 'Active',
  COMPLETED: 'Completed',
  CANCELLED: 'Cancelled',
};

/** Approval workflow state (isApproved) → display label. */
export const APPROVAL_STATUS_LABEL = {
  1: 'Approved',
  2: 'Pending',
  3: 'On hold',
  4: 'Completed',
};

export function fromGrantListResponse(dto) {
  return {
    ...dto,
    fundClassLabel: dto.fundClassCode ? `Class ${dto.fundClassCode}` : '—',
    statusLabel: GRANT_STATUS_LABEL[dto.status] || (dto.isActive ? 'Active' : 'Inactive'),
    approvalStatusLabel: APPROVAL_STATUS_LABEL[dto.isApproved] || '—',
  };
}

export function fromGrantDetailsResponse(dto) {
  return fromGrantListResponse(dto);
}

/** GrantDetailsResponse → edit-form values (strings, mirroring grantFormDefaults). */
export function toGrantFormValues(grant) {
  return {
    grantCode: grant.grantCode || '',
    donorId: grant.donorId != null ? String(grant.donorId) : '',
    fundProfileId: grant.fundProfileId != null ? String(grant.fundProfileId) : '',
    agreementName: grant.agreementName || '',
    status: grant.status || 'ACTIVE',
    agreementDate: grant.agreementDate || '',
    startDate: grant.startDate || '',
    endDate: grant.endDate || '',
    approvalStatus: grant.isApproved != null ? String(grant.isApproved) : '2',
    approvedBy: grant.approvedBy != null ? String(grant.approvedBy) : '',
    // approvalDate is a timestamp on the wire; the form field is a date.
    approvalDate: grant.approvalDate ? String(grant.approvalDate).slice(0, 10) : '',
    approvalRemarks: grant.approvalRemarks || '',
    description: grant.description || '',
    agreementDocumentPath: grant.agreementDocumentPath || '',
  };
}

/**
 * Form values → CreateGrantRequest. Donor, class and the total (Σ of the fund
 * profile's tranche plan) are all derived server-side from fundProfileId.
 */
export function toCreateGrantRequest(values) {
  return {
    // Omitted on create → backend auto-generates ZRY/GA/YYYY/NNN.
    grantCode: values.grantCode?.trim() || undefined,
    fundProfileId: Number(values.fundProfileId),
    agreementName: values.agreementName.trim(),
    status: values.status || 'ACTIVE',
    agreementDate: values.agreementDate,
    startDate: values.startDate,
    endDate: values.endDate,
    approvalStatus: Number(values.approvalStatus || 2),
    approvedBy: values.approvedBy ? Number(values.approvedBy) : null,
    approvalDate: values.approvalDate || null,
    approvalRemarks: values.approvalRemarks?.trim() || null,
    description: values.description?.trim() || null,
    agreementDocumentPath: values.agreementDocumentPath?.trim() || null,
  };
}
